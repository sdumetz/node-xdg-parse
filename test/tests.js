'use strict';
var path = require("path");
const fs = require("fs");
const {readFile} = fs.promises;

const {parse, serialize} = require("../index.js")

let cases = fs.readdirSync(path.resolve(__dirname, "fixtures"));


function load(file){
  return readFile(path.resolve(__dirname,"fixtures", file), {encoding: "utf8"});
}

describe("parse : ",function(){

  it("throw on invalid parameter", function(){
    expect(()=>{ parse(null)}).to.throw(Error);
    expect(()=>{ parse({})}).to.throw(Error);
  })
  it("throw on invalid line", function(){
    expect(()=>{ parse("hello world")}).to.throw(Error);
  })

  it("parse empty string", function(){
    expect(parse("")).to.deep.equal({});
  })

  it("split multiple line breaks", function(){
    expect(parse(`[FOO]\n\nbar=hello world`)).to.deep.equal({FOO:{bar:"hello world"}});
  })

  it("split CRLF lines (duh!)", function(){
    expect(parse(`[FOO]\r\nbar=hello world`)).to.deep.equal({FOO:{bar:"hello world"}});
  })

  it("parse boolean types", function(){
    expect(parse(`[FOO]\nbar=true`)).to.deep.equal({FOO:{bar:true}});
  })

  it("can parse empty string values", function(){
    expect(parse(`[FOO]\nbar=\n`)).to.deep.equal({FOO:{bar:""}});
  })

  it("can parse a locale", function(){
    return load(path.resolve(__dirname, "fixtures", "index.theme"))
    .then((data)=>{
      expect(parse(data, "sv")).to.matchSnapshot();
      expect(parse(data, "fr")).to.matchSnapshot();
    })
  })
  cases.forEach(function(file){
    it(`parse(${file})`, function(){
      return load(file).then(function(data){
        expect(parse(data)).to.matchSnapshot();
      });
    })
  });

})

describe("serialize : ",function(){
  cases.forEach(function(file){
    it(`serialize ${file}`, function(){
      return load(file)
      .then(d => parse(d))
      .then(function(data){
        const output_data = parse(serialize(data));
        expect(output_data).to.deep.equal(data);
      });
    });
  });
})


