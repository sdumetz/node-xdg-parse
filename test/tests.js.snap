// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse :  can parse a locale 1`] = `
Object {
  "32x32/apps": Object {
    "Context": "Applications",
    "Size": 32,
    "Type": "Fixed",
  },
  "48x48/apps": Object {
    "Context": "Applications",
    "Size": 48,
    "Type": "Fixed",
  },
  "48x48/mimetypes": Object {
    "Context": "MimeTypes",
    "Size": 48,
    "Type": "Fixed",
  },
  "Icon Theme": Object {
    "Comment": "Träinspirerat ikontema",
    "Directories": "48x48/apps,48x48/mimetypes,32x32/apps,scalable/apps,scalable/mimetypes",
    "Inherits": "wood,default",
    "Name": "Björk",
  },
  "scalable/apps": Object {
    "Context": "Applications",
    "MaxSize": 256,
    "MinSize": 1,
    "Size": 48,
    "Type": "Scalable",
  },
  "scalable/mimetypes": Object {
    "Context": "MimeTypes",
    "MaxSize": 256,
    "MinSize": 1,
    "Size": 48,
    "Type": "Scalable",
  },
}
`;

exports[`parse :  can parse a locale 2`] = `
Object {
  "32x32/apps": Object {
    "Context": "Applications",
    "Size": 32,
    "Type": "Fixed",
  },
  "48x48/apps": Object {
    "Context": "Applications",
    "Size": 48,
    "Type": "Fixed",
  },
  "48x48/mimetypes": Object {
    "Context": "MimeTypes",
    "Size": 48,
    "Type": "Fixed",
  },
  "Icon Theme": Object {
    "Comment": "theme d'icones avec textures bois",
    "Directories": "48x48/apps,48x48/mimetypes,32x32/apps,scalable/apps,scalable/mimetypes",
    "Inherits": "wood,default",
    "Name": "Birch",
  },
  "scalable/apps": Object {
    "Context": "Applications",
    "MaxSize": 256,
    "MinSize": 1,
    "Size": 48,
    "Type": "Scalable",
  },
  "scalable/mimetypes": Object {
    "Context": "MimeTypes",
    "MaxSize": 256,
    "MinSize": 1,
    "Size": 48,
    "Type": "Scalable",
  },
}
`;

exports[`parse :  parse(comments.desktop) 1`] = `
Object {
  "Section": Object {
    "key": "Value",
  },
}
`;

exports[`parse :  parse(crlf.desktop) 1`] = `
Object {
  "Desktop Entry": Object {
    "Comment": "Files with CRLF line endings are the most fun",
    "Name": "CRLF File",
    "Type": "Application",
    "Version": "1.0",
  },
}
`;

exports[`parse :  parse(fooview.desktop) 1`] = `
Object {
  "Desktop Action Create": Object {
    "Exec": "fooview --create-new",
    "Icon": "fooview-new",
    "Name": "Create a new Foo!",
  },
  "Desktop Action Gallery": Object {
    "Exec": "fooview --gallery",
    "Name": "Browse Gallery",
  },
  "Desktop Entry": Object {
    "Actions": "Gallery;Create",
    "Comment": "The best viewer for Foo objects available!",
    "Exec": "fooview %F",
    "Icon": "fooview",
    "MimeType": "image/x-foo",
    "Name": "Foo Viewer",
    "TryExec": "fooview",
    "Type": "Application",
    "Version": "1.0",
  },
}
`;

exports[`parse :  parse(index.theme) 1`] = `
Object {
  "32x32/apps": Object {
    "Context": "Applications",
    "Size": 32,
    "Type": "Fixed",
  },
  "48x48/apps": Object {
    "Context": "Applications",
    "Size": 48,
    "Type": "Fixed",
  },
  "48x48/mimetypes": Object {
    "Context": "MimeTypes",
    "Size": 48,
    "Type": "Fixed",
  },
  "Icon Theme": Object {
    "Comment": "Icon theme with a wooden look",
    "Directories": "48x48/apps,48x48/mimetypes,32x32/apps,scalable/apps,scalable/mimetypes",
    "Inherits": "wood,default",
    "Name": "Birch",
  },
  "scalable/apps": Object {
    "Context": "Applications",
    "MaxSize": 256,
    "MinSize": 1,
    "Size": 48,
    "Type": "Scalable",
  },
  "scalable/mimetypes": Object {
    "Context": "MimeTypes",
    "MaxSize": 256,
    "MinSize": 1,
    "Size": 48,
    "Type": "Scalable",
  },
}
`;

exports[`parse :  parse(mimeapps.list) 1`] = `
Object {
  "Added Associations": Object {
    "application/bar": "foo",
    "text/plain": "test.desktop",
  },
  "Default Applications": Object {
    "application/baz": "foobar.desktop",
    "text/plain": "test.desktop",
  },
}
`;
